<a href="https://colab.research.google.com/github/nagusubra/Solar_panel_dust_detection/blob/main/Final_report.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Install libraries and modules

In [1]:
!pip install -q tensorflow-model-optimization
!pip install openpyxl
!pip install xlsxwriter

import xlsxwriter
import openpyxl
from openpyxl import Workbook
from openpyxl.drawing.image import Image

import tempfile
from tensorflow import keras
import tensorflow_model_optimization as tfmot

from tqdm import tqdm_notebook as tqdm
import os
import time
import pickle
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 240.6/240.6 kB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 78.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.12.0 requires numpy<1.24,>=1.22, but you have numpy 1.24.2 which is incompatible.
numba 0.56.4 requires numpy<1.24,>=1.18, but you have numpy 1.24.2 which is incompatible.
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 152.8/152.8 kB 5.5 MB/s eta 0:00:00


In [2]:


import tensorflow as tf
from tensorflow.keras.models import Model, load_model
from tensorflow.keras.layers import Input, Conv2D, MaxPooling2D, Dense, Flatten, BatchNormalization, Dropout
from tensorflow.keras.preprocessing import image_dataset_from_directory
from tensorflow.keras.optimizers import Adam, SGD
from tensorflow.keras.utils import plot_model
from tensorflow.keras.losses import binary_crossentropy
import tensorflow_datasets as tfds
import zipfile

In [3]:
# mounting google drive (if you are using Colab)
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Evaluation function

In [4]:
# Evaluate Model Size
def get_gzipped_model_size(file):
  # Returns size of gzipped model, in bytes.

  _, zipped_file = tempfile.mkstemp('.zip')
  with zipfile.ZipFile(zipped_file, 'w', compression=zipfile.ZIP_DEFLATED) as f:
    f.write(file)
  return os.path.getsize(zipped_file)


def evaluate_model(model_path, model_info, val_dataset):

  # Evaluate test accuracy and test loss
  model = tf.keras.models.load_model(model_path)
  test_loss, test_acc = model.evaluate(val_dataset, verbose=0)

  # Evaluate Model Size
  model_size = get_gzipped_model_size(model_path)

  # Evaluate Inference Time
  startTime = time.time()
  prediction = model.predict(val_dataset)
  executionTime = (time.time() - startTime)/len(val_dataset)

  # Print
  print('\nModel Accuracy:', test_acc*100, '%')
  print("Model Size: %.2f bytes" % (model_size))
  print("Inference Time is: ", executionTime, "s")

  # Build Evalution dataframe
  evulation_dict = {
                      "Evaluation type": "Evualation",
                      "Model Information": model_info,
                      "Accuracy": str(test_acc*100) + " %",
                      "Loss": str(test_loss*100) + " %",
                      "Model Size": str(model_size) + " bytes",
                      "Inference Time": str(executionTime) + " sec"
                    }
  
  evulation_df = pd.DataFrame.from_dict(evulation_dict, orient='index').reset_index()


  return test_acc, model_size, executionTime, evulation_df


def evaluate_model_without_saving_stats(model, model_path, model_info, val_dataset):

  # Evaluate test accuracy and test loss
  # model = tf.keras.models.load_model(model_path)
  test_loss, test_acc = model.evaluate(val_dataset, verbose=0)

  # Evaluate Model Size
  model_size = get_gzipped_model_size(model_path)

  # Evaluate Inference Time
  startTime = time.time()
  prediction = model.predict(val_dataset)
  executionTime = (time.time() - startTime)/len(val_dataset)

  # Print
  print('\nModel Accuracy:', test_acc*100, '%')
  print("Model Size: %.2f bytes" % (model_size))
  print("Inference Time is: ", executionTime, "s")

  # Build Evalution dataframe
  evulation_dict = {
                      "Evaluation type": "Evualation",
                      "Model Information": model_info,
                      "Accuracy": str(test_acc*100) + " %",
                      "Loss": str(test_loss*100) + " %",
                      "Model Size": str(float(model_size)/10**6) + " Mega bytes", # we want mega bytes
                      "Inference Time": str(executionTime) + " sec"
                    }
  
  evulation_df = pd.DataFrame.from_dict(evulation_dict, orient='index').reset_index()


  return test_acc, model_size, executionTime, evulation_df




# research paper evaluation function
def evaluate_research_paper_with_model_path(model_path):
  # model_path = "/content/models/solnet.hdf5"
  solnet = load_model(model_path, compile=False)
  history = solnet.history()
  plt.plot(history.history['loss'])
  plt.plot(history.history['acc'])
  plt.title('acc loss vs epoch')
  plt.xlabel('epoch')
  plt.legend(['loss', 'acc'], loc='upper left')
  plt.show()

def evaluate_research_paper_with_model(model):
  history = model.history()
  plt.plot(history.history['loss'])
  plt.plot(history.history['acc'])
  plt.title('acc loss vs epoch')
  plt.xlabel('epoch')
  plt.legend(['loss', 'acc'], loc='upper left')
  plt.show()

In [5]:
def iterative_pruning(model, initial_sparsity, final_sparsity, begin_step, end_step, tdata_loader,vdata_loader, epochs):
  prune_low_magnitude = tfmot.sparsity.keras.prune_low_magnitude

  # Define model for pruning.
  pruning_params = {
      'pruning_schedule': tfmot.sparsity.keras.PolynomialDecay(initial_sparsity=initial_sparsity,
          final_sparsity=final_sparsity, begin_step=begin_step, end_step=end_step, frequency=100)
  }

  pruned_model = prune_low_magnitude(model, **pruning_params)

  # `prune_low_magnitude` requires a recompile.
  optimizer = tf.keras.optimizers.Adam(learning_rate=1e-5)
  pruned_model.compile(optimizer=Adam(.0001, .8, .9),
      loss=binary_crossentropy,
      metrics=['accuracy'])

  callbacks = [
      tfmot.sparsity.keras.UpdatePruningStep(),
  ]

  pruned_model.fit(tdata_loader, epochs=epochs, validation_data=vdata_loader, callbacks=callbacks)

  # Strip pruning wrappers
  stripped_pruned_model = tfmot.sparsity.keras.strip_pruning(pruned_model)

  return pruned_model, stripped_pruned_model

In [6]:
def print_model_weights_sparsity(model):

    for layer in model.layers:
        if isinstance(layer, tf.keras.layers.Wrapper):
            weights = layer.trainable_weights
        else:
            weights = layer.weights
        for weight in weights:
            if "kernel" not in weight.name or "centroid" in weight.name:
                continue
            weight_size = weight.numpy().size
            zero_num = np.count_nonzero(weight == 0)
            print(
                f"{weight.name}: {zero_num/weight_size:.2%} sparsity ",
                f"({zero_num}/{weight_size})",
            )

# Load data

In [7]:
batch_size = 32
#location = 'dataset/'
location = "/content/drive/MyDrive/Solar_panel_dust_detection/dataset_1"
label_mode = 'binary'
seed = 10 #changed for each fold made manually

class_names = ['clean', 'dirty']
in_size = [227, 227, 3]

tr_dataset = image_dataset_from_directory(directory=location, label_mode= label_mode, class_names=class_names,
                                          seed=seed, labels='inferred', image_size=in_size[:-1], 
                                          subset = 'training', batch_size=batch_size, validation_split=.2)

val_dataset = image_dataset_from_directory(directory=location, label_mode= label_mode, class_names=class_names,
                                          seed=seed, labels='inferred', image_size=in_size[:-1],
                                          subset = 'validation', batch_size=batch_size, validation_split=.2)

Found 1440 files belonging to 2 classes.
Using 1152 files for training.
Found 1440 files belonging to 2 classes.
Using 288 files for validation.


# Load research paper model and Evaluate

In [ ]:
research_paper_model_path = '/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/research_paper_model.h5'
research_paper_model = tf.keras.models.load_model(research_paper_model_path)

In [ ]:
research_model_test_acc, research_model_model_size, research_model_executionTime, research_model_evulation_df = evaluate_model_without_saving_stats(research_paper_model, research_paper_model_path, "#0", val_dataset)

9/9 [==============================] - 40s 1s/step

Model Accuracy: 52.77777910232544 %
Model Size: 212498787.00 bytes
Inference Time is:  4.588925785488552 s


In [ ]:
research_model_evulation_df

index                      0
0    Evaluation type             Evualation
1  Model Information                     #0
2           Accuracy    52.77777910232544 %
3               Loss   1045.3107833862305 %
4         Model Size  212.498787 Mega bytes
5     Inference Time  4.588925785488552 sec

# Load base model and Evaluate


In [8]:
base_model_path = '/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/final_base_model.h5'
base_model = tf.keras.models.load_model(base_model_path)

In [ ]:
base_model_test_acc, base_model_model_size, base_model_executionTime, base_model_evulation_df = evaluate_model_without_saving_stats(base_model, base_model_path, "#0", val_dataset)

9/9 [==============================] - 28s 41ms/step

Model Accuracy: 82.63888955116272 %
Model Size: 647985790.00 bytes
Inference Time is:  4.551086213853624 s


In [ ]:
base_model_evulation_df

index                      0
0    Evaluation type             Evualation
1  Model Information                     #0
2           Accuracy    82.63888955116272 %
3               Loss    90.77435731887817 %
4         Model Size   647.98579 Mega bytes
5     Inference Time  4.551086213853624 sec

# Load iteratively pruned model and Evaluate

In [ ]:
# Loading pruned models gives an error, hence we will iteratively prune it in the notebook

# Iteratively prune model and Evaluate

In [9]:
# frquency = 100
# target sparsity = 80%
# sparisty for each pruning tep = 75%/100 = 0.8%
# Total Number of Training Samples = 1440
# batch size = 32
# epochs = 10

# possible start and end for the iterative pruning schedule
# Number of learning steps = (1440/32)*10 = 450 (45, 400)
# Number of learning steps = (1440/32)*20 = 900 (90, 800)
# Number of learning steps = (1440/32)*25 = 1125 (110, 1010) -> chosen epoch of 25 epochs
# Number of learning steps = (1440/32)*30 = 1350 (135, 1200)
# Number of learning steps = (1440/32)*35 = 1575 (150, 1400)

# iterative_pruning(
                  #     model, 
                  #     initial_sparsity = 0, 
                  #     final_sparsity   = 0.7, 
                  #     begin_step       = 0, 110  (~10% of 5625) # pruning early on the pruning schedule to get optimal model size
                  #     end_step         = ?, 1010 (~80% of 5625) # pruning only till 80% so that model can still have enough leanring steps to learn and build networks for the data
                  #     train_images, 
                  #     train_labels, 
                  #     epochs           = 3, no change ? since the Number of Steps per Epoch = (Total Number of Training Samples = ?) / (Batch Size = ?) 
                  # )

iteratively_pruned_model, iteratively_pruned_stripped_model = iterative_pruning(base_model, 0, 0.8, 110, 1010, tr_dataset, val_dataset, 25)

Epoch 1/25
36/36 [==============================] - 327s 6s/step - loss: 0.9867 - accuracy: 0.8003 - val_loss: 4.4123 - val_accuracy: 0.6042
Epoch 2/25
36/36 [==============================] - 157s 4s/step - loss: 0.5022 - accuracy: 0.8889 - val_loss: 1.8994 - val_accuracy: 0.7708
Epoch 3/25
36/36 [==============================] - 120s 3s/step - loss: 0.5699 - accuracy: 0.8976 - val_loss: 7.7062 - val_accuracy: 0.5799
Epoch 4/25
36/36 [==============================] - 139s 3s/step - loss: 0.5196 - accuracy: 0.9262 - val_loss: 2.0001 - val_accuracy: 0.6667
Epoch 5/25
36/36 [==============================] - 120s 3s/step - loss: 0.3014 - accuracy: 0.9392 - val_loss: 9.2650 - val_accuracy: 0.5799
Epoch 6/25
36/36 [==============================] - 121s 3s/step - loss: 0.3469 - accuracy: 0.9418 - val_loss: 3.6080 - val_accuracy: 0.5972
Epoch 7/25
36/36 [==============================] - 121s 3s/step - loss: 0.1574 - accuracy: 0.9575 - val_loss: 0.3490 - val_accuracy: 0.9201
Epoch 8/25
36

In [10]:
iteratively_pruned_model_path = "iteratively_pruned_model.h5"
iteratively_pruned_stripped_model.save(iteratively_pruned_model_path)


iteratively_pruned_model.save('/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/iteratively_pruned_80_model.h5')
iteratively_pruned_stripped_model.save('/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/iteratively_stripped_pruned_80_model.h5')

In [44]:
# tf.keras.saving.save_model(
#     iteratively_pruned_model, 
#     '/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/iteratively_pruned_75_model_new_method.h5', 
#     overwrite=True, 
#     save_format="h5"
# )


# tf.keras.saving.save_model(
#     iteratively_pruned_stripped_model, 
#     '/content/drive/MyDrive/Solar_panel_dust_detection/final_base_model/iteratively_stripped_pruned_75_model_new_method.h5', 
#     overwrite=True, 
#     save_format="h5"
# )

In [11]:
pruned_test_acc, pruned_model_size, pruned_executionTime, pruned_evulation_df = evaluate_model_without_saving_stats(iteratively_pruned_model, iteratively_pruned_model_path, "#1", val_dataset)

9/9 [==============================] - 24s 38ms/step

Model Accuracy: 68.05555820465088 %
Model Size: 68465710.00 bytes
Inference Time is:  2.706649992201063 s


In [12]:
print_model_weights_sparsity(iteratively_pruned_model)

conv2d/kernel:0: 79.12% sparsity  (27572/34848)
conv2d_1/kernel:0: 79.12% sparsity  (486126/614400)
conv2d_2/kernel:0: 79.12% sparsity  (700022/884736)
conv2d_3/kernel:0: 79.12% sparsity  (1050032/1327104)
conv2d_4/kernel:0: 79.12% sparsity  (700022/884736)
dense/kernel:0: 79.12% sparsity  (29867588/37748736)
dense_1/kernel:0: 79.12% sparsity  (13274482/16777216)
dense_2/kernel:0: 79.13% sparsity  (3241/4096)


In [13]:
pruned_evulation_df

index                      0
0    Evaluation type             Evualation
1  Model Information                     #1
2           Accuracy    68.05555820465088 %
3               Loss   166.39598608016968 %
4         Model Size    68.46571 Mega bytes
5     Inference Time  2.706649992201063 sec

## Observation

We have almost 86% decrease in model size with minimal loss in accuracy.



| Base model | Iteratively pruned model|
|------------|------------|
| - Overall accuracy: 82.63%| - Overall accuracy: 82.63% (higher accuracy)|
| - Overall model size: 647.98 MB | - Overall model size: 89.66 MB (smaller size)|
| - Overall Inference time: 4.55 s | - Overall Inference time: 3.28 s (lower inference time)|
| - Performance: Lower| - Performance: Higher|
| - Key: It has a larger model size compared to iterative pruning,| - Key: Iterative pruning proves that we can fine tune our sparsity level, |
|  and significanlty has higher inference time.| while maintaining our accuracy and inference time.|

In [ ]:
pruned_evulation_df

index                      0
0    Evaluation type             Evualation
1  Model Information                     #1
2           Accuracy    94.09722089767456 %
3               Loss   54.101359844207764 %
4         Model Size         89648959 bytes
5     Inference Time  9.117764472961426 sec

# Quantization

In [18]:
# A helper function to evaluate the TF Lite model using "test" dataset.
def evaluate_model_2(interpreter, model_path, val_dataset):
  input_index = interpreter.get_input_details()[0]["index"]
  output_index = interpreter.get_output_details()[0]["index"]

  input_details = interpreter.get_input_details()

  # Run predictions on every image in the "test" dataset.
  prediction_digits = []
  # for test_image in test_images:
  for images, labels in val_dataset.unbatch().batch(1):
    # Pre-processing: add batch dimension and convert to float32 to match with
    # the model's input data format.
    # test_image = np.expand_dims(test_image, axis=0).astype(np.float32)
    interpreter.set_tensor(input_details[0]['index'], images)

    # interpreter.set_tensor(input_index, images)

    # Run inference.
    startTime = time.time()
    interpreter.invoke()
    executionTime = (time.time() - startTime)/len(images)

    # Post-processing: remove batch dimension and find the digit with highest
    # probability.
    output = interpreter.tensor(output_index)
    digit = np.argmax(output()[0])
    prediction_digits.append(digit)

  # Compare prediction results with ground truth labels to calculate accuracy.
  accurate_count = 0
  for index in range(len(prediction_digits)):
    if prediction_digits[index] == labels:
      accurate_count += 1
  accuracy = accurate_count * 1.0 / len(prediction_digits)

  model_size = get_gzipped_model_size(model_path)
  # Print
  print('\nModel Accuracy:', accuracy*100, '%')
  print("Model Size: %.2f bytes" % (model_size))
  print("Inference Time is: ", executionTime, "s")


  # Build Evalution dataframe
  evulation_dict = {
                      "Evaluation type": "Evualation",
                      "Accuracy": str(accuracy*100) + " %",
                      "Model Size": str(float(model_size)/10**6) + " Mega bytes", # we want mega bytes
                      "Inference Time": str(executionTime) + " sec"
                    }
  
  evulation_df = pd.DataFrame.from_dict(evulation_dict, orient='index').reset_index()


  return evulation_df

In [19]:
# A helper function to evaluate the TF Lite model using a test loader.
def evaluate_tflite_model(interpreter, model_path, test_loader):
  input_index = interpreter.get_input_details()[0]["index"]
  output_index = interpreter.get_output_details()[0]["index"]

  # Run predictions on every image in the test dataset.
  prediction_digits = []
  accurate_count = 0
  total_images = 0
  total_time = 0

  for images, labels in val_dataset:
    total_images += len(images)
    for i in range(len(images)):
      # Pre-processing: add batch dimension and convert to float32 to match with
      # the model's input data format.
      test_image = np.expand_dims(images[i].numpy(), axis=0).astype(np.float32)
      interpreter.set_tensor(input_index, test_image)

      # Run inference.
      startTime = time.time()
      interpreter.invoke()
      executionTime = time.time() - startTime
      total_time += executionTime

      # Post-processing: remove batch dimension and find the digit with highest
      # probability.
      output = interpreter.tensor(output_index)
      digit = np.argmax(output()[0])
      prediction_digits.append(digit)

      # Compare prediction results with ground truth labels to calculate accuracy.
      if digit == labels[i].numpy():
        accurate_count += 1

  accuracy = accurate_count * 1.0 / total_images
  model_size = get_gzipped_model_size(model_path)

  # Print
  print('\nModel Accuracy:', accuracy*100, '%')
  print("Model Size: %.2f bytes" % (model_size))
  print("Average Inference Time per Image: %.5f s" % (total_time/total_images))

  # Build Evalution dataframe
  evulation_dict = {
                      "Evaluation type": "Evualation",
                      "Accuracy": str(accuracy*100) + " %",
                      "Model Size": str(float(model_size)/10**6) + " Mega bytes", # we want mega bytes
                      "Average Inference Time per Image": str((total_time/total_images)) + " sec"
                    }
  
  evulation_df = pd.DataFrame.from_dict(evulation_dict, orient='index').reset_index()
  
  return evulation_df

In [40]:
# A helper function to evaluate the TF Lite model using a test loader.
def evaluate_tflite_model_2(interpreter, model_path, test_loader):
  input_index = interpreter.get_input_details()[0]["index"]
  output_index = interpreter.get_output_details()[0]["index"]

  # Run predictions on every image in the test dataset.
  prediction_digits = []
  accurate_count = 0
  total_images = 0
  total_time = 0

  total_images_2 = 0

  for images, labels in val_dataset:
    total_images_2 = len(images)

    total_images += len(images)
    for i in range(len(images)):
      # Pre-processing: add batch dimension and convert to float32 to match with
      # the model's input data format.
      test_image = np.expand_dims(images[i].numpy(), axis=0).astype(np.float32)
      interpreter.set_tensor(input_index, test_image)
      # interpreter.set_tensor(input_index, images)


      # Run inference.
      startTime = time.time()
      interpreter.invoke()
      executionTime = time.time() - startTime
      total_time += executionTime

      # Post-processing: remove batch dimension and find the digit with highest
      # probability.
      output = interpreter.tensor(output_index)
      digit = np.argmax(output()[0])
      prediction_digits.append(digit)

      # Compare prediction results with ground truth labels to calculate accuracy.
      test_label = np.expand_dims(labels[i].numpy(), axis=0).astype(np.float32)
      if digit == test_label:
      # if digit == labels[i].numpy():
        accurate_count += 1

  accuracy = accurate_count * 1.0 / total_images
  accuracy_2 = accurate_count * 1.0 / total_images_2
  model_size = get_gzipped_model_size(model_path)

  # Print
  print('\nModel Accuracy:', accuracy*100, '%')
  print('\nModel Accuracy:', accuracy_2*100, '%')
  print("Model Size: %.2f bytes" % (model_size))
  print("Average Inference Time per Image: %.5f s" % (total_time/total_images))

  # Build Evalution dataframe
  evulation_dict = {
                      "Evaluation type": "Evualation",
                      "Accuracy": str(accuracy*100) + " %",
                      "Accuracy_2": str(accuracy_2*100) + " %",
                    
                      "Model Size": str(float(model_size)/10**6) + " Mega bytes", # we want mega bytes
                      "Average Inference Time per Image": str((total_time/total_images)) + " sec"
                    }
  
  evulation_df = pd.DataFrame.from_dict(evulation_dict, orient='index').reset_index()
  
  return evulation_df

In [20]:
converter = tf.lite.TFLiteConverter.from_keras_model(base_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_types = [tf.float16]
fp16_quantized_tflite_base_model = converter.convert()

In [21]:
converter = tf.lite.TFLiteConverter.from_keras_model(iteratively_pruned_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_types = [tf.float16]
fp16_quantized_tflite_iteratively_pruned_model = converter.convert()

In [22]:
with open('fp16_quantized_tflite_base_model.tflite', 'wb') as f:
  f.write(fp16_quantized_tflite_base_model)
interpreter_fp16_quant_base_model = tf.lite.Interpreter(model_path=str('fp16_quantized_tflite_base_model.tflite'))
interpreter_fp16_quant_base_model.allocate_tensors()

In [23]:
with open('fp16_quantized_tflite_iteratively_pruned_model.tflite', 'wb') as f:
  f.write(fp16_quantized_tflite_iteratively_pruned_model)
interpreter_fp16_quant_iteratively_pruned_model = tf.lite.Interpreter(model_path=str('fp16_quantized_tflite_iteratively_pruned_model.tflite'))
interpreter_fp16_quant_iteratively_pruned_model.allocate_tensors()

In [24]:
input_data = val_dataset

In [25]:
quantized_tflite_base_model_evulation_df = evaluate_model_2(interpreter_fp16_quant_base_model, 'fp16_quantized_tflite_base_model.tflite', val_dataset)


Model Accuracy: 100.0 %
Model Size: 39712494.00 bytes
Inference Time is:  0.08251786231994629 s


In [ ]:
quantized_tflite_base_model_evulation_df

index                        0
0  Evaluation type               Evualation
1         Accuracy                  100.0 %
2       Model Size     50.241397 Mega bytes
3   Inference Time  0.07050800323486328 sec

In [ ]:
quantized_tflite_iteratively_pruned_model_evulation_df = evaluate_model_2(interpreter_fp16_quant_iteratively_pruned_model, 'fp16_quantized_tflite_iteratively_pruned_model.tflite', val_dataset)


Model Accuracy: 100.0 %
Model Size: 61294937.00 bytes
Inference Time is:  0.3843374252319336 s


In [ ]:
quantized_tflite_iteratively_pruned_model_evulation_df

index                       0
0  Evaluation type              Evualation
1         Accuracy                 100.0 %
2       Model Size    61.294937 Mega bytes
3   Inference Time  0.3843374252319336 sec

In [26]:
quantized_tflite_base_model_evulation_df = evaluate_tflite_model(interpreter_fp16_quant_base_model, 'fp16_quantized_tflite_base_model.tflite', val_dataset)


Model Accuracy: 52.77777777777778 %
Model Size: 39712494.00 bytes
Average Inference Time per Image: 0.08116 s


In [27]:
quantized_tflite_base_model_evulation_df

index                       0
0                   Evaluation type              Evualation
1                          Accuracy     52.77777777777778 %
2                        Model Size    39.712494 Mega bytes
3  Average Inference Time per Image  0.0811604575978385 sec

In [28]:
quantized_tflite_iteratively_pruned_model_evulation_df = evaluate_tflite_model(interpreter_fp16_quant_iteratively_pruned_model, 'fp16_quantized_tflite_iteratively_pruned_model.tflite', val_dataset)


Model Accuracy: 52.77777777777778 %
Model Size: 49448854.00 bytes
Average Inference Time per Image: 0.41845 s


In [29]:
quantized_tflite_iteratively_pruned_model_evulation_df

index                       0
0                   Evaluation type              Evualation
1                          Accuracy     52.77777777777778 %
2                        Model Size    49.448854 Mega bytes
3  Average Inference Time per Image  0.4184533730149269 sec

In [41]:
quantized_tflite_iteratively_pruned_model_evulation_df_2 = evaluate_tflite_model_2(interpreter_fp16_quant_iteratively_pruned_model, 'fp16_quantized_tflite_iteratively_pruned_model.tflite', val_dataset)


Model Accuracy: 52.77777777777778 %

Model Accuracy: 475.0 %
Model Size: 49448854.00 bytes
Average Inference Time per Image: 0.41853 s


In [43]:
quantized_tflite_iteratively_pruned_model_evulation_df_2

index                       0
0                   Evaluation type              Evualation
1                          Accuracy     52.77777777777778 %
2                        Accuracy_2                 475.0 %
3                        Model Size    49.448854 Mega bytes
4  Average Inference Time per Image  0.4185345371564229 sec

## 8 bit quantization

In [30]:
converter = tf.lite.TFLiteConverter.from_keras_model(base_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
# converter.target_spec.supported_types = [tf.float16]
fp8_quantized_tflite_base_model = converter.convert()

with open('fp8_quantized_tflite_base_model.tflite', 'wb') as f:
  f.write(fp8_quantized_tflite_base_model)
interpreter_fp8_quant_base_model = tf.lite.Interpreter(model_path=str('fp8_quantized_tflite_base_model.tflite'))
interpreter_fp8_quant_base_model.allocate_tensors()



converter = tf.lite.TFLiteConverter.from_keras_model(iteratively_pruned_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
# converter.target_spec.supported_types = [tf.float16]
fp8_quantized_tflite_iteratively_pruned_model = converter.convert()

with open('fp8_quantized_tflite_iteratively_pruned_model.tflite', 'wb') as f:
  f.write(fp8_quantized_tflite_iteratively_pruned_model)
interpreter_fp8_quant_iteratively_pruned_model = tf.lite.Interpreter(model_path=str('fp8_quantized_tflite_iteratively_pruned_model.tflite'))
interpreter_fp8_quant_iteratively_pruned_model.allocate_tensors()

In [31]:
quantized_8_bit_tflite_base_model_evulation_df = evaluate_tflite_model(interpreter_fp8_quant_base_model, 'fp8_quantized_tflite_base_model.tflite', val_dataset)
quantized_8_bit_tflite_base_model_evulation_df


Model Accuracy: 52.77777777777778 %
Model Size: 19146399.00 bytes
Average Inference Time per Image: 0.06352 s


index                        0
0                   Evaluation type               Evualation
1                          Accuracy      52.77777777777778 %
2                        Model Size     19.146399 Mega bytes
3  Average Inference Time per Image  0.06351853327618705 sec

In [32]:
quantized_8_bit_tflite_iteratively_pruned_model_evulation_df = evaluate_tflite_model(interpreter_fp8_quant_iteratively_pruned_model, 'fp8_quantized_tflite_iteratively_pruned_model.tflite', val_dataset)
quantized_8_bit_tflite_iteratively_pruned_model_evulation_df


Model Accuracy: 52.77777777777778 %
Model Size: 79725259.00 bytes
Average Inference Time per Image: 0.43853 s


index                        0
0                   Evaluation type               Evualation
1                          Accuracy      52.77777777777778 %
2                        Model Size     79.725259 Mega bytes
3  Average Inference Time per Image  0.43853385829263264 sec

## Observation

